71 boot sequence update

kryczkal opened the pull request:
This is a massive refactor / enhancement that introduces chainloading functionality into the OS.
Things done in this PR:
- Refactored the entire `alkos/kernel/arch/x86_64` directory
- Split x86_64 into boot32 and boot64
- Boot32 became a loader elf that sets up the 32 bit mode and then jumps to actual kernel
- Boot32 is now compiled with i386 toolchain and us such is an called as external project from the cmake
- Files in `x86_64/common` are shared between Boot32 and Boot64
- Added module loading functionality to the make iso script and cmake
- Rewrote Boot32 PreKernelInit in C
- Added simple parsing of Elf files (to load kernel into memory and commence long jump to it)
- Reworked download script to install both i386 toolchain and x86_64
- Boot32 passes data to Boot64 using a LoaderData structure

kryczkal commented:
This is a big PR, that achieves the desired functionality, but overall is still VERY rough around the edges and requires cleanup. On the other hand the fact that it is working made me request reviews.
Initial comments are appreciated.

kryczkal commented:
BUG: I broke the dependencies and running "Run" doesn't run target alkos.kernel and libk automatically.

kryczkal commented:
![image](https://github.com/user-attachments/assets/dad8cec3-8f8c-41d3-9cd7-0b2e32398722)
PASS FAIL PASS FAIL and infinite loop is not something I like to see

kryczkal commented:
> BUG: I broke the dependencies and running "Run" doesn't run target alkos.kernel and libk automatically.

Fixed

kryczkal commented on alkos/kernel/arch/x86_64/CMakeLists.txt:
```diff
@@ -1,93 +1,89 @@
 
-set(SRTI_FILE "${CMAKE_CURRENT_SOURCE_DIR}/cxx/crti.nasm")
-set(SRTN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/cxx/crtn.nasm")
+if (NOT KERNEL_MODULES)
```
This is a mistake, this could be empty

kryczkal commented on alkos/kernel/arch/x86_64/boot32/boot32.nasm:
```diff
@@ -0,0 +1,37 @@
+          extern PreKernelInit
+
+          ; GDT64
+          extern GDT64.Pointer
```
this can be deleted

kryczkal commented on alkos/kernel/arch/x86_64/boot32/boot32.nasm:
```diff
@@ -0,0 +1,37 @@
+          bits 32
+
+          MULTI_BOOT_INFO_T_LOCATION equ stack_top - 4
```
this can be deleted

kryczkal commented on alkos/kernel/arch/x86_64/boot32/src/enter_kernel.nasm:
```diff
@@ -0,0 +1,33 @@
+          jmp GDT64.Code:.jmp_kernel
+.jmp_kernel:
+          mov edi, [ebp + 12] ; LoaderData address
+          mov eax, [k_ptr]
```
this is essentialy a hack, and i ought to give reference to what it does and where i got it from

kryczkal commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+
+    TerminalWriteString(INFO_TAG "Parsing Multiboot2 tags...\n");
+
+    multiboot_tag_module* kernel_module = nullptr;
```
This function is too long

Jlisowskyy replied:
I think loader should be moved out from here and it will look good then

kryczkal replied:
Added as TODO

kryczkal commented on alkos/kernel/arch/x86_64/common/abi/panic.cpp:
```diff
@@ -0,0 +1,25 @@
+    TerminalWriteError(msg);
+    TerminalWriteError("\n");
+
+    // #TODO
```
Fixing this is a TODO yes

kryczkal commented on alkos/kernel/arch/x86_64/common/include/comp.h:
```diff
@@ -7,28 +7,28 @@
 /**
-* @WARNING: FUNCTIONALITY IS NOT WORKING YET DUE TO SOME PROBLEMS WITH LINKING AND OFFESTING CALLS BY 4 BYTES EACH TIME
-*/
+ * @WARNING: FUNCTIONALITY IS NOT WORKING YET DUE TO SOME PROBLEMS WITH LINKING AND OFFESTING CALLS
```
This whole file should be deleted

Jlisowskyy replied:
Yesss

kryczkal commented:
@Jlisowskyy  tell me if you have more comments than I have already given myself. And let's look into why the tests fail.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/definitions/page_tables.nasm:
```diff
@@ -0,0 +1,17 @@
+          bits 32
+
+          global p4_table
+          global p3_table
+          global p2_table
+          global p1_table
+
+          section   .bss
+          align 4096
+p4_table:
+          resb 4096 ; Page Map Level 4
+p3_table:
+          resb 4096 ; Page Directory Pointer Table
+p2_table:
+          resb 4096 ; Page Directory
+p1_table:
+          resb 4096 ; Page Table
```
Shouldnt it be defined in C code instead?

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/elf.hpp:
```diff
@@ -0,0 +1,72 @@
+//
+// Created by wookie on 1/4/25.
+//
+
```
Should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/elf.hpp:
```diff
@@ -0,0 +1,72 @@
+
+// ELF-64 Structures
+
+#define ELF_MAGIC0 0x7F
+#define ELF_MAGIC1 'E'
+#define ELF_MAGIC2 'L'
+#define ELF_MAGIC3 'F'
```
static constexpr

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/libc.hpp:
```diff
@@ -0,0 +1,40 @@
+//
+// Created by wookie on 1/4/25.
+//
```
Should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/multiboot2.h:
```diff
@@ -0,0 +1,483 @@
+//
+// Created by wookie on 1/5/25.
+//
```
Should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/tags.hpp:
```diff
@@ -0,0 +1,12 @@
+//
+// Created by wookie on 1/4/25.
+//
```
Should be removed.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/libc.hpp:
```diff
@@ -0,0 +1,40 @@
+// Created by wookie on 1/4/25.
+//
+
+#ifndef ALK_OS_KERNEL_ARCH_X86_64_BOOT32_INCLUDE_LIBC_HPP
+#define ALK_OS_KERNEL_ARCH_X86_64_BOOT32_INCLUDE_LIBC_HPP
```
Incorrect header guard. Refer to Google C++ style guide.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/tags.hpp:
```diff
@@ -0,0 +1,12 @@
+// Created by wookie on 1/4/25.
+//
+
+#ifndef ALK_OS_KERNEL_ARCH_X86_64_BOOT32_INCLUDE_TAGS_HPP
+#define ALK_OS_KERNEL_ARCH_X86_64_BOOT32_INCLUDE_TAGS_HPP
```
Incorrect header guard. Refer to Google C++ style guide.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/linker.ld:
```diff
@@ -0,0 +1,48 @@
+          /* https://www.gnu.org/software/grub/manual/multiboot2/multiboot.html#OS-image-format */
+          /* " The Multiboot2 header must be contained completely within the first 32768 bytes of the OS image. " */
+	. = 2k;
+	PROVIDE(multiboot_header_start = .);
+
+          .multiboot ALIGN(64) : AT(.)
+          {
+                    KEEP(*(.multiboot))
+	}
+
+          PROVIDE(multiboot_header_end = .);
```
Possibly wrong line alignment.

kryczkal replied:
this is probably a github bug. It's ok in the code

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/elf.cpp:
```diff
@@ -0,0 +1,85 @@
+//
+// Created by wookie on 1/4/25.
+//
```
Shoule be removed.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/elf.cpp:
Missing newline

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
```diff
@@ -0,0 +1,49 @@
+//
+// Created by wookie on 1/4/25.
+//
```
Should be removed.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
```diff
@@ -0,0 +1,49 @@
+#include <stdint.h>
+
+// Extern Declarations for Page Tables defined in Assembly
+extern uint64_t p4_table[];
+extern uint64_t p3_table[];
+extern uint64_t p2_table[];
+extern uint64_t p1_table[];
```
Why we can't define tables here?

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
```diff
@@ -0,0 +1,49 @@
+extern uint64_t p1_table[];
+
+// Page Table Entry Flags
+#define PRESENT_BIT       (1 << 0)  // The page is present in memory
+#define WRITE_BIT         (1 << 1)  // The page is writable
+#define USER_BIT          (1 << 2)  // The page is accessible from user mode
+#define WRITE_THROUGH_BIT (1 << 3)  // Writes go directly to memory
+#define CACHE_DISABLE_BIT (1 << 4)  // The page is not cached
+#define ACCESSED_BIT      (1 << 5)  // CPU sets this bit when the page is accessed
+#define DIRTY_BIT         (1 << 6)  // CPU sets this bit when the page is written to
+#define HUGE_PAGE_BIT     (1 << 7)  // Creates a 2 MiB page in PD table
+#define GLOBAL_BIT        (1 << 8)  // The page isn't flushed from the TLB on address space switch
```
static constexpr

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
Missing newline

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
```diff
@@ -0,0 +1,49 @@
+    // Identity map the first 1 GiB of memory using 2 MiB pages
+    for (int i = 0; i < 512; i++) {
+        // Each PD entry maps a 2 MiB page
+        p2_table[i] = ((uint64_t)(0x200000 * i)) | PRESENT_BIT | WRITE_BIT | HUGE_PAGE_BIT;
```
random number -> use variable instead of comment

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/enables/setup_page_tables.cpp:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @return int 0 because the function always succeeds
+ */
+int setup_page_tables()
```
Function names should be PascalCase.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+//
+// Created by wookie on 1/2/25.
+//
```
should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+extern "C" int check_cpuid();
+extern "C" int check_long_mode();
+
+void setup_page_tables();
```
Possibly setup header or introduce extern keyword here?

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+// Data structure that holds information passed from the 32-bit loader to the 64-bit kernel
+LoaderData loader_data;
+
+extern "C" void PreKernelInit(uint32_t boot_loader_magic, void* multiboot_info_addr)
```
Add comments to the function to divide it into sections and enhance readability

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+extern "C" void enter_kernel(void* kernel_entry, void* multiboot_info_addr);
+
+// Buffer for text output
+char text_buffer[1024];
```
Could be moved to stack instead of global memory.

kryczkal replied:
Doesn't matter, the loader will be yeeted out from memory anyway. And the stack is a dangling piece of memory anyway after the loader passes control and boot64 establishes it's own stack. It has 0 significance.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/init.cpp:
```diff
@@ -0,0 +1,167 @@
+
+    multiboot_tag_module* kernel_module = nullptr;
+    multiboot_tag* tag;
+    for (tag = (multiboot_tag*)(multiboot_info_addr + 8);
+         tag->type != MULTIBOOT_TAG_TYPE_END;
+         tag = (multiboot_tag*)((multiboot_uint8_t*)tag + ((tag->size + 7) & ~7))) {
+        TerminalWriteString(INFO_TAG "Found tag: ");
+        TerminalWriteString(GetTagName(tag->type));
+        TerminalWriteString("\n");
+
+        if(tag->type == MULTIBOOT_TAG_TYPE_MODULE) {
+            multiboot_tag_module* module = (multiboot_tag_module*)tag;
+            kernel_module = module;
+
+            // TODO: Modules should be loaded with a specific command line
+            // so that the loader knows which module is the kernel
+            // but for now, we will just take the last module as the kernel
+
+            TerminalWriteString(INFO_TAG "Module loaded at: ");
+            uint32_to_string(module->mod_start, text_buffer);
+            TerminalWriteString(text_buffer);
+            TerminalWriteString(" - ");
+            uint32_to_string(module->mod_end, text_buffer);
+            TerminalWriteString(text_buffer);
+            TerminalWriteString("\n");
+            TerminalWriteString(INFO_TAG "With command line: ");
+            if ( (char*)module->cmdline == "\0") {
+                TerminalWriteString((char*)module->cmdline);
+            } else {
+                TerminalWriteString("no command line provided");
+            }
+            TerminalWriteString("\n");
```
Loading could be moved to separate function.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/include/libc.hpp:
```diff
@@ -0,0 +1,40 @@
+ * @param value The value to convert
+ * @param buffer The buffer to store the string
+ */
+void uint32_to_string(uint32_t value, char* buffer);
```
I am not sure if stanard libc contains such function. I would prefer to use PascalCase for our own functions.

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot32/src/libc.cpp:
```diff
@@ -0,0 +1,52 @@
+//
+// Created by wookie on 1/4/25.
+//
```
should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/boot64/definitions/gdt.nasm:
```diff
@@ -0,0 +1,47 @@
+          RW        equ 1 << 1 ; Readable/Writable
+          AC        equ 1 << 0 ; Accessed
+          ; Flags bits
+          GRAN_4K   equ 1 << 7 ; 4 KiB granularity
+          SZ_32     equ 1 << 6 ; 32-bit segment
+          LONG_MODE equ 1 << 5 ; Long mode
+
+          section   .data
+          align 16
+GDT64:
+          .Null:
+          dq 0
+          .Code: equ $ - GDT64
+          dw 0x0000
+          dw 0x0000
+          db 0x00
+          db 0x9A        ; Access byte
+          db 0x20        ; Flags
+          db 0x00
+          .Data: equ $ - GDT64
+          dw 0x0000
+          dw 0x0000
+          db 0x00
+          db 0x92        ; Access byte
+          db 0x00        ; Flags
+          db 0x00
+.End:
+          .Pointer:
+          dw .End - GDT64 - 1 ;
+          dq GDT64
+
+          global GDT64
+          global GDT64.Pointer
+          global GDT64.Code
+          global GDT64.Data
+          global kKernelCodeOffset
+          global kKernelDataOffset
+
+          kKernelCodeOffset dd GDT64.Code
+          kKernelDataOffset dd GDT64.Data
```
It should be rewritten to C, but I would not do it in this PR.

kryczkal replied:
I created an issue

Jlisowskyy commented on alkos/kernel/arch/x86_64/common/include/loader_data.hpp:
```diff
@@ -0,0 +1,31 @@
+//
+// Created by wookie on 1/5/25.
+//
```
Should be removed

Jlisowskyy commented on alkos/kernel/arch/x86_64/common/include/loader_data.hpp:
```diff
@@ -0,0 +1,31 @@
+// Created by wookie on 1/5/25.
+//
+
+#ifndef ALK_OS_KERNEL_ARCH_X86_64_COMMON_INCLUDE_LOADER_DATA_H
```
Incorrect header guard.

Jlisowskyy commented on alkos/toolchains/x86_64-toolchain.cmake:
```diff
@@ -7,7 +7,7 @@ set(CMAKE_SYSTEM_PROCESSOR x86_64)
 
 # Define the path to your cross-compiler tools
-set(CROSS_COMPILER_PATH "${CMAKE_CURRENT_LIST_DIR}/../../tools/bin")
+set(CROSS_COMPILER_PATH "${CMAKE_CURRENT_LIST_DIR}/../../tools/x86_64-elf/bin")
```
Use prefix here.

Jlisowskyy commented on alkos/toolchains/i386-toolchain.cmake:
```diff
@@ -0,0 +1,31 @@
+set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
+
+# Define the path to your cross-compiler tools
+set(CROSS_COMPILER_PATH "${CMAKE_CURRENT_LIST_DIR}/../../tools/i386-elf/bin")
```
Use prefix here

Jlisowskyy reviewed (changes requested):
Small changes

kryczkal commented:
![image](https://github.com/user-attachments/assets/a4779615-d31a-40b8-996a-9f6844519918)
Addressed comments and fixed tests.
Unfortunatelly had to use defines in a lot of places.

Jlisowskyy reviewed (approved):
Works beatifully
